linux.lisp

```diff
@@ -56,23 +56,67 @@
+
+(defconstant +s-ifmt+ #o170000)
+(defconstant +s-iflnk+ #o120000)
+(defun s-istype (mode mask) (= (logand mode +s-ifmt+) mask))
+(defun s-islnk (mode) (s-istype mode +s-iflnk+))
+
+(defun symlinkp (path)
   (cffi:with-foreign-objects ((stat '(:struct stat)))
-    (when (< (cffi:foreign-funcall "stat" :string (pathname-utils:native-namestring path) :pointer stat :int) 0)
+    (when (< (cffi:foreign-funcall "lstat" :string (pathname-utils:native-namestring path) :pointer stat :int) 0)
       (fail (cffi:foreign-funcall "strerror" :int64 errno)))
-    (let ((dev (stat-dev stat)))
-      (do-proc ((l :int) (r :int) (name :char 32))
-          ("/proc/self/mountinfo" "%*d %*d %d:%d / %*[^-]- %*s /dev/%s"
-                                  (fail "Device not found in mountinfo table"))
-        (when (and (= l (ldb (byte 32 8) dev))
-                   (= r (1- (ldb (byte 8 0) dev))))
-          (return (cffi:foreign-string-to-lisp name :max-chars 32)))))))
+    (s-islnk (stat-mode stat))))
+
+(defun pathname-force-file (path)
+  (cond
+    ((pathname-utils:root-p path) path)
+    ((pathname-utils:file-p path) path)
+    (t (let ((directories (pathname-directory path)))
+         (make-pathname :defaults path
+                        :directory (butlast directories)
+                        :name (car (last directories)))))))
+
+(defun find-mount-root (path)
+  (labels ((dev-id (path)
+             (cffi:with-foreign-objects ((stat '(:struct stat)))
+               (when (< (cffi:foreign-funcall "stat" :string (pathname-utils:native-namestring path) :pointer stat :int) 0)
+                 (fail (cffi:foreign-funcall "strerror" :int64 errno)))
+               (stat-dev stat)))
+           (rec (path &optional (id (dev-id path)))
+             (if (pathname-utils:root-p path)
+                 path
+                 (let* ((parent (pathname-utils:parent path))
+                        (parent-id (dev-id parent)))
+                   (if (= parent-id id)
+                       (rec parent parent-id)
+                       path)))))
+    (let* ((root (rec path))
+           (root-file (pathname-force-file root)))
+      (if (symlinkp root-file)
+          (find-mount-root (readlink root-file))
+          root))))
+
+(define-implementation storage-device (path)
+  (let* ((mount-root (pathname-utils:native-namestring (pathname-force-file (find-mount-root path)))))
+    (do-proc ((mountpoint :char 512) (name :char 32))
+        ("/proc/self/mountinfo" "%*d %*d %*d:%*d / %s %*[^-]- %*s %s"
+                                (fail "Device not found in mountinfo table"))
+      (when (= 0 (cffi:foreign-funcall "strncmp" :pointer mountpoint :string mount-root :size 32 :int))
+        (return (cffi:foreign-string-to-lisp name :max-chars 32))))))
 
 (define-implementation storage-device-path (device)
   (do-proc ((mount :char 512) (name :char 32))
-      ("/proc/self/mountinfo" "%*d %*d %*d:%*d / %s %*s %*s - %*s /dev/%s"
```

Comment:
This will *not* work. The device name is *not* a path, and functions like storage-io-bytes expect a device name, not a path.

Reply:
I've removed it again in main.